src/qutip_cupy/cupy_dense.py

```diff
@@ -0,0 +1,88 @@
+import cupy as cp
+import numbers
+from qutip.core import data
+
+
+class CuPyDense(data.Data):
+    def __init__(self, data, shape=None, copy=True):
+        base = cp.array(data, dtype=cp.complex128, order='K', copy=copy)
+        if shape is None:
+            shape = base.shape
+            # Promote to a ket by default if passed 1D data.
+            if len(shape) == 1:
+                shape = (shape[0], 1)
+        if not (
+            len(shape) == 2
+            and isinstance(shape[0], numbers.Integral)
+            and isinstance(shape[1], numbers.Integral)
+            and shape[0] > 0
+            and shape[1] > 0
+        ):
+            raise ValueError("shape must be a 2-tuple of positive ints, but is " + repr(shape))
+        if shape and (shape[0] != base.shape[0] or shape[1] != base.shape[1]):
+            if shape[0] * shape[1] != base.size:
+                raise ValueError("".join([
+                    "invalid shape ",
+                    str(shape),
+                    " for input data with size ",
+                    str(base.size)
+                ]))
+            else:
+                self._cp = base.reshape(shape)
+        else:
+
+            self._cp = base
+
+        super().__init__((shape[0], shape[1]))
+
+    def copy(self):
+        return CuPyDense(self._cp.copy())
+
+    def to_array(self):
+        """ 
```

Comment:
This docstring can be removed, I think. As discussed in the call today `help(CuPyDense.to_array)` will read the docstring from `Data.to_array`.

Reply:
In this case `Data.to_array` does not have any doctring,and for that matter most of the methods of `data.Data` since they are not "implemented" .  Besides the behaviour is different here, we are not only copying to a numpy array  we are migrationg  the data from the GPU to the CPU which is a bigger penalty and the user should be made aware.